crates/sc-proof-of-time/src/clock_master.rs

```diff
@@ -0,0 +1,211 @@
+//! Clock master implementation.
+
+use crate::pot_state::PotProtocolState;
+use crate::utils::{get_consensus_tip_proofs, topic, PotGossip, LOG_TARGET};
+use crate::{BootstrapParams, PotComponents};
+use futures::{FutureExt, StreamExt};
+use parity_scale_codec::{Decode, Encode};
+use sc_network::PeerId;
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedSender};
+use sp_blockchain::{HeaderBackend, Info};
+use sp_consensus::SyncOracle;
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+use std::thread;
+use std::time::Instant;
+use subspace_core_primitives::{NonEmptyVec, PotProof, PotSeed};
+use subspace_proof_of_time::ProofOfTime;
+use tracing::{debug, error, info, warn};
+
+/// The clock master manages the protocol: periodic proof generation/verification, gossip.
+pub struct ClockMaster<Block: BlockT, Client, SO> {
+    proof_of_time: Arc<ProofOfTime>,
+    gossip: PotGossip<Block>,
+    client: Arc<Client>,
+    sync_oracle: Arc<SO>,
+    pot_state: Arc<dyn PotProtocolState>,
+    chain_info_fn: Arc<dyn Fn() -> Info<Block> + Send + Sync>,
+}
+
+impl<Block, Client, SO> ClockMaster<Block, Client, SO>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block>,
+    SO: SyncOracle + Send + Sync + Clone + 'static,
+{
+    /// Creates the clock master instance.
+    pub fn new(
+        components: PotComponents,
+        gossip: PotGossip<Block>,
+        client: Arc<Client>,
+        sync_oracle: Arc<SO>,
+        chain_info_fn: Arc<dyn Fn() -> Info<Block> + Send + Sync>,
+    ) -> Self {
+        let PotComponents {
+            proof_of_time,
+            protocol_state: pot_state,
+            ..
+        } = components;
+
+        Self {
+            proof_of_time,
+            pot_state,
+            gossip,
+            client,
+            sync_oracle,
+            chain_info_fn,
+        }
+    }
+
+    /// Starts the workers.
+    pub async fn run(self, bootstrap_params: Option<BootstrapParams>) {
+        if let Some(params) = bootstrap_params.as_ref() {
+            // The clock master is responsible for bootstrapping, build/add the
+            // initial proof to the state and start the proof producer.
+            self.add_bootstrap_proof(params);
```

Comment:
The state needs to be created ahead of time (during new_partial()) so it can be shared with consensus/block import. It currently structured so that the component initializes the first proof when run() starts. I think you are asking why this can't be done before hand during new_partial()

Yes, this is possible. I can do this as part of testing with consensus integration

Reply:
No, I think what I'm asking is why `InternalState.chain` is `Vec<PotProof>` and not a vector of inputs to proofs. I don't really understand why `self.add_bootstrap_proof` generates a proof internally, proofs are only supposed to be generated in a dedicated thread where proof of time is evaluated.

And it is still a bit confusing to me when you mention `new_partial` because it is not in the code we're reviewing.

Reply:
As we discussed today, this part still has some gaps. This will be fixed in the next change where we do consensus/bootstrap integration